SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp

```diff
@@ -0,0 +1,387 @@
+						   && AuthorityComponent->AuthorityOnBeginPlay == AuthorityComponent->AuthorityOnTick
+						   && AuthorityComponent->NumAuthorityGains == 1 && AuthorityComponent->NumAuthorityLosses == 0;
+				}
+				else
+				{
+					bIsStateValid = AuthorityComponent->ReplicatedAuthorityOnBeginPlay != 0 && AuthorityComponent->AuthorityOnBeginPlay == 0
+									&& AuthorityComponent->AuthorityOnTick == 0 && AuthorityComponent->NumAuthorityGains == 0
+									&& AuthorityComponent->NumAuthorityLosses == 0;
+				}
+
+				if(bIsStateValid)
+				{
+					FinishStep();
+				}
+			}, 5.0f);
+	}
+
+	// GameState.
+	{
+		AddStep(
+			TEXT("GameState - Determine Authority by every Server"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				// Again, no need to wait for more time.
+				ASpatialAuthorityTestGameState* GameState = GetWorld()->GetGameState<ASpatialAuthorityTestGameState>();
+				if (GameState == nullptr) // GameMode already checked on previous step.
+				{
+					FinishTest(EFunctionalTestResult::Failed, TEXT("This test requires ASpatialAuthorityTestGameState"));
+					return;
+				}
+
+				bool bIsStateValid;
+
+				USpatialAuthorityTestActorComponent* AuthorityComponent = GameState->AuthorityComponent;
+				// Either it's bigger than 0 and all match the GameMode Authority (in the Authoritive Server),
+				// or all equal to zero (except for replicated).
+				if (AuthorityComponent->AuthorityOnBeginPlay > 0)
+				{
+					ASpatialAuthorityTestGameMode* GameMode = GetWorld()->GetAuthGameMode<ASpatialAuthorityTestGameMode>();
+					int GameModeAuthority = GameMode->AuthorityComponent->ReplicatedAuthorityOnBeginPlay;
+					bIsStateValid = AuthorityComponent->AuthorityOnBeginPlay == GameModeAuthority
+						   && AuthorityComponent->ReplicatedAuthorityOnBeginPlay == GameModeAuthority
+						   && AuthorityComponent->AuthorityOnTick == GameModeAuthority
+						   && AuthorityComponent->NumAuthorityGains == 1
+						   && AuthorityComponent->NumAuthorityLosses == 0;
+				}
+				else
+				{
+
+					bIsStateValid = AuthorityComponent->ReplicatedAuthorityOnBeginPlay != 0
+						   && AuthorityComponent->AuthorityOnBeginPlay == 0
+						   && AuthorityComponent->AuthorityOnTick == 0
+						   && AuthorityComponent->NumAuthorityGains == 0
+						   && AuthorityComponent->NumAuthorityLosses == 0;
+				}
+
+				if(bIsStateValid)
+				{
+					FinishStep();
+				}
+			},
+			5.0f);
+	}
+
+}
+
+void ASpatialAuthorityTest::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ASpatialAuthorityTest, DynamicReplicatedActor);
+}
+
+void ASpatialAuthorityTest::CrossServerSetDynamicReplicatedActor_Implementation(ASpatialAuthorityTestReplicatedActor* Actor)
+{
+	DynamicReplicatedActor = Actor;
+}
+
+bool ASpatialAuthorityTest::VerifyTestActor(ASpatialAuthorityTestActor* Actor, int AuthorityOnBeginPlay, int AuthorityOnTick
```

Comment:
This could just take an AActor, and dynamically get the AuthorityComponent. That way we could use it for the GameState and GameMode checks also.

Reply:
The reason it wasn't done is because all the other ones check X == Y, but when doing the GameMode / GameState they also have a check of ReplicatedAuthWorkerIdOnBeginPlay != 0. I can merge them, and just do some detection if it's gamemode / gamestate to check differently for them, but since I was doing that case separately I just put them in their own step.